src/modules/mc_land_detector/mc_land_detector_main.cpp

```diff
@@ -0,0 +1,199 @@
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file mc_land_detector_main.cpp
+ * Land detection algorithm for multicopters
+ *
+ * @author Johan Jansen <jnsn.johan@gmail.com>
+ */
+
+#include <unistd.h>					//usleep
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <drivers/drv_hrt.h>
+#include <systemlib/systemlib.h>	//Scheduler
+#include <systemlib/err.h>			//print to console
+
+#include "MulticopterLandDetector.h"
+
+//Function prototypes
+static int mc_land_detector_start();
+static void mc_land_detector_stop();
+
+/**
+ * land detector app start / stop handling function
+ * This makes the land detector module accessible from the nuttx shell
+ * @ingroup apps
+ */
+extern "C" __EXPORT int mc_land_detector_main(int argc, char *argv[]);
+
+//Private variables
+static MulticopterLandDetector *mc_land_detector_task = nullptr;
+static int _landDetectorTaskID = -1;
+
+/**
+* Deamon thread function
+**/
+static void mc_land_detector_deamon_thread(int argc, char *argv[])
+{
+	mc_land_detector_task->landDetectorLoop();
+}
+
+/**
+* Stop the task, force killing it if it doesn't stop by itself
+**/
+static void mc_land_detector_stop()
+{
+	if (mc_land_detector_task == nullptr || _landDetectorTaskID == -1) {
+		errx(1, "not running");
+		return;
+	}
+
+	mc_land_detector_task->shutdown();
+
+	//Wait for task to die
+	int i = 0;
+
+	do {
+		/* wait 20ms */
+		usleep(20000);
+
+		/* if we have given up, kill it */
+		if (++i > 50) {
+			task_delete(_landDetectorTaskID);
+			break;
+		}
+	} while (mc_land_detector_task->isRunning());
+
+
+	delete mc_land_detector_task;
+	mc_land_detector_task = nullptr;
+	_landDetectorTaskID = -1;
+	warn("mc_land_detector has been stopped");
+}
+
+/**
+* Start new task, fails if it is already running. Returns OK if successful
+**/
+static int mc_land_detector_start()
+{
+	if (mc_land_detector_task != nullptr || _landDetectorTaskID != -1) {
+		errx(1, "already running");
+		return -1;
+	}
+
+	//Allocate memory
+	mc_land_detector_task = new MulticopterLandDetector();
+
+	if (mc_land_detector_task == nullptr) {
+		errx(1, "alloc failed");
+		return -1;
+	}
+
+	//Start new thread task
+	_landDetectorTaskID = task_spawn_cmd("mc_land_detector",
+					     SCHED_DEFAULT,
+					     SCHED_PRIORITY_DEFAULT,
+					     1024,
+					     (main_t)&mc_land_detector_deamon_thread,
+					     nullptr);
+
+	if (_landDetectorTaskID < 0) {
+		errx(1, "task start failed: %d", -errno);
+		return -1;
+	}
+
+	/* avoid memory fragmentation by not exiting start handler until the task has fully started */
```

Comment:
/* avoid memory fragmentation by not exiting start handler until the task has fully started */
@LorenzMeier Is this logic really needed to avoid memory fragmentation?

I'll take a look at what it takes to extract the generic parts of task spawning, handling and termination into a syslib. My "boilerplate" code assumes there will only be one instance of each running task.